### PR TITLE
Add endpoint to get items on a list

### PR DESCRIPTION
**Pull request by samstradling**
And attempted to write some tests for it.

### CODE REVIEW

File: db/db.go [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/dynamodb"
+	"github.com/aws/aws-sdk-go/service/dynamodb/dynamodbiface"
+	"github.com/mount-joy/thelist-lambda/data"
+)
+
+// DB - interface for talking to the database
+type DB interface {
+	GetItemsOnList(listID *string) (*[]data.Item, error)
```
Comment (alexNaylorTeece) [inline db/db.go:13]: I think this is just a taste thing, but I prefer it when the interface doesn't state the name of the variable(s)

File: db/db.go [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+}
+
+type db struct {
```
[the remark this answers is not in the document]
Comment (alexNaylorTeece) [reply db/db.go:16]: Could also put all the dynamo specific stuff in a file called `dynamo.go` and the generic interface, and anything else that's generic in `db.go`

File: db/getItemsOnList.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
+package db
+
+import (
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/dynamodb"
+	"github.com/aws/aws-sdk-go/service/dynamodb/dynamodbattribute"
+	"github.com/mount-joy/thelist-lambda/data"
+)
+
+func (d *db) GetItemsOnList(listID *string) (*[]data.Item, error) {
```
Comment (alexNaylorTeece) [inline db/getItemsOnList.go:10]: A whole file for a single function feels like we will end up with hundreds of files

How would you feel about us putting all dynamo db call functions in a single file?

File: db/db.go (hunk 1 above)
Comment (alexNaylorTeece) [reply db/db.go:13]: No clear why this would take a pointer to string, and return a pointer to item

File: db/getItemsOnList.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+		KeyConditionExpression: aws.String("ListId = :id"),
+		TableName:              aws.String(dbTableNameItems),
+	}
+
+	result, err := d.session.Query(input)
+	if err != nil {
+		return nil, err
+	}
+	if result.Items == nil {
+		return nil, nil
```
Comment (alexNaylorTeece) [inline db/getItemsOnList.go:24]: For a `Get` where the got thing is `nil`, should we not be returning a not found error?

File: db/getItemsOnList.go [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+	}
+
+	items := []data.Item{}
+
```
Comment (alexNaylorTeece) [inline db/getItemsOnList.go:28]: Not clear what this space is for

File: db/getItemsOnList.go [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+	for _, i := range result.Items {
+		item := new(data.Item)
+		err = dynamodbattribute.UnmarshalMap(i, &item)
+		if err != nil {
+			return nil, err
```
Comment (alexNaylorTeece) [inline db/getItemsOnList.go:33]: If we have managed to get through some items before we errored, would there be value in returning the half complete list that was got successfully?

File: db/getItemsOnList_test.go [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+				{
+					Item: "Apples",
+					ID:   "bb0d5e8e",
+				},
+			},
+			expectedErr: nil,
+		},
+		{
+			name:        "When Query returns an error, that error is returned",
+			output:      nil,
```
Comment (alexNaylorTeece) [inline db/getItemsOnList_test.go:71]: Would be good to have something valid returned from the mock here so that you are definitely testing that you don't return what was in the result if the err != nil

File: db/getItemsOnList_test.go [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+			outputErr:   errors.New("Something went wrong"),
+			expectedRes: nil,
+			expectedErr: errors.New("Something went wrong"),
+		},
+	}
```
Comment (alexNaylorTeece) [inline db/getItemsOnList_test.go:76]: Can you add a test case for `output=nil` and `outputerr=nil`

File: db/getItemsOnList_test.go [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			dbMocked := &mockDB{}
+			dbMocked.Test(t)
+			defer dbMocked.AssertExpectations(t)
+
+			listID := "474c2Fff7"
```
Comment (alexNaylorTeece) [inline db/getItemsOnList_test.go:84]: If this is same for every test case, could define it outside the loop
Or could just hard code it in line 87

File: handlers/getItems.go [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+	"github.com/mount-joy/thelist-lambda/db"
+)
+
+type getItems struct {
+	db db.DB
+}
+
+func newGetItems() RouteHandler {
+	return &getItems{
+		db: db.NewDB(),
```
Comment (alexNaylorTeece) [inline handlers/getItems.go:22]: Given that DB will I assume eventually do more than GetItems, the naming of some of this might need thinking about, otherwise we are going to end up with a `db.NewDB()` on a `deleteItems` a `putItems` etc.

File: handlers/getItems.go [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+	}
+}
+
+func (g *getItems) match(request events.APIGatewayProxyRequest) bool {
+	var re = regexp.MustCompile(`^/list/([\w-]+)/items/?$`)
```
Comment (alexNaylorTeece) [inline handlers/getItems.go:27]: Can you add a test added that checks this regex

File: handlers/getItems.go [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+
+	if err != nil {
+		log.Printf("Error: %s", err.Error())
+		return nil, http.StatusInternalServerError
+	}
+
+	return items, http.StatusOK
+}
+
+func getListID(path *string) (*string, error) {
```
Comment (alexNaylorTeece) [inline handlers/getItems.go:51]: Tests for this would be helpful I think

**Review by alexNaylorTeece — DISMISSED** [review]

File: db/db.go (hunk 1 above)
Comment (samstradling) [reply db/db.go:13]: > No clear why this would take a pointer to string, and return a pointer to item

As in you you're asking why it isn't:

```suggestion
	GetItemsOnList(listID string) ([]data.Item, error)
```

File: db/getItemsOnList.go (hunk 3 above)
Comment (samstradling) [reply db/getItemsOnList.go]: I'm not convinced by this, wouldn't lots of ~40 line files be better than one ~320 line one (assuming about 8 endpoints)?

File: db/getItemsOnList.go (hunk 6 above)
Comment (samstradling) [reply db/getItemsOnList.go:34]: I'm not sure there is, especially as it'd be pretty catastrophic if this was failing, right?

File: handlers/getItems.go (hunk 10 above)
Comment (samstradling) [reply handlers/getItems.go]: I think this is sorted 👍

File: handlers/getItems.go (hunk 11 above)
Comment (samstradling) [reply handlers/getItems.go]: You're just too eager

File: db/constants.go [hunk 13]
```diff
@@ -0,0 +1,5 @@
+package db
+
+const dbEndpoint string = "http://192.168.1.153:8000/"
```
Comment (Joseph-Cashmore) [inline db/constants.go:3]: We spoke about this in meeting on 23/10 - will need to find a way of avoiding using our IP

**Review by Joseph-Cashmore — APPROVED** [review]

File: Makefile [hunk 14]
```diff
@@ -4,7 +4,7 @@ build:
 	sam build
 
 start: build
-	sam local start-api
+	sam local start-api --docker-network host
```
Comment (samstradling) [inline Makefile:7]: This is required for the lambda to be able to talk to a locally running db
